File: wenger-feature/src/main/java/io/github/rcarlosdasilva/wenger/feature/context/EnvironmentHandler.java
package io.github.rcarlosdasilva.wenger.feature.context;

import com.google.common.base.Enums;
import com.google.common.base.Optional;
import com.google.common.base.Strings;
import io.github.rcarlosdasilva.kits.sys.SystemHelper;
import io.github.rcarlosdasilva.wenger.common.constant.GeneralConstant;
import io.github.rcarlosdasilva.wenger.feature.config.AppProperties;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.config.BeanPostProcessor;
import org.springframework.boot.context.ContextIdApplicationContextInitializer;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.core.env.Environment;
import org.springframework.stereotype.Component;

import java.util.Objects;
import java.util.stream.Stream;

/**
 * 环境
 *
 * @author <a href="mailto:rcarlosdasilva@qq.com">Dean Zhao</a>
 */
@Slf4j
@Component
@EnableConfigurationProperties({AppProperties.class})
public class EnvironmentHandler extends ContextIdApplicationContextInitializer implements BeanPostProcessor {

  @Autowired
  private Environment env;
  @Autowired
  private AppProperties appProperties;

  @Getter
  private RuntimeProfile runtimeProfile;
  @Getter
  private String applicationName;
  @Getter
  private String tempDir;

  @Override
  public void initialize(ConfigurableApplicationContext applicationContext) {
    applicationName = applicationContext.getApplicationName();
    if (Strings.isNullOrEmpty(applicationName)) {
      applicationName = GeneralConstant.DEFAULT_APPLICATION_NAME;
    }
    super.initialize(applicationContext);
  }

  @Override
  public Object postProcessBeforeInitialization(Object bean, String beanName) {
    String[] profiles = env.getActiveProfiles();
    if (profiles.length > 1) {
      log.warn("[Environment] - 激活了多个Profile，执行环境将会是第一个激活的预置环境，建议使用单个Profile");
    }
    Stream.of(profiles).forEach(profile -> {
      if (runtimeProfile != null) {
        return;
      }

      Optional<RuntimeProfile> orf = Enums.getIfPresent(RuntimeProfile.class, profile.toUpperCase());
      if (orf.isPresent()) {
        runtimeProfile = orf.get();
      }
    });

    if (runtimeProfile == null) {
      log.warn("[Environment] - 未找到预置环境（devel, ci, test, prepro, production）");
      runtimeProfile = RuntimeProfile.CUSTOM_DEFINED;
    }

    this.tempDir = Objects.toString(appProperties.getTempDir(), SystemHelper.tempDir());
    log.info("[Environment] - 使用临时目录：{}", tempDir);

    EnvironmentHandlerHolder.setEnvironmentHandler(this);
    return bean;
  }

  public static class EnvironmentHandlerHolder {

    private static EnvironmentHandler environmentHandler;

    private static void setEnvironmentHandler(EnvironmentHandler environmentHandler) {
      EnvironmentHandlerHolder.environmentHandler = environmentHandler;
    }

    public static EnvironmentHandler get() {
      return environmentHandler;
    }

    private EnvironmentHandlerHolder() {
      throw new IllegalStateException();
    }

  }

}

File: wenger-feature/src/main/java/io/github/rcarlosdasilva/wenger/feature/context/EnvironmentHandler.kt
package io.github.rcarlosdasilva.wenger.feature.context

import com.google.common.base.Enums
import io.github.rcarlosdasilva.kits.sys.SystemHelper
import io.github.rcarlosdasilva.wenger.common.constant.GeneralConstant
import io.github.rcarlosdasilva.wenger.feature.config.AppProperties
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.config.BeanPostProcessor
import org.springframework.boot.context.ContextIdApplicationContextInitializer
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.core.env.Environment
import org.springframework.stereotype.Component

/**
 * 环境
 *
 * @author [Dean Zhao](mailto:rcarlosdasilva@qq.com)
 */
@Component
@EnableConfigurationProperties(value = [AppProperties::class])
class EnvironmentHandler @Autowired constructor(
  private val env: Environment,
  private val appProperties: AppProperties
) : ContextIdApplicationContextInitializer(), BeanPostProcessor {

  private val logger = KotlinLogging.logger {}

  lateinit var runtimeProfile: RuntimeProfile
  lateinit var applicationName: String
  lateinit var tempDir: String

  override fun initialize(applicationContext: ConfigurableApplicationContext) {
    applicationName = applicationContext.applicationName ?: GeneralConstant.DEFAULT_APPLICATION_NAME
    super.initialize(applicationContext)
  }

  override fun postProcessBeforeInitialization(bean: Any, beanName: String?): Any? {
    val profiles = env.activeProfiles
    if (profiles.size > 1) {
      logger.warn { "[Environment] - 激活了多个Profile，执行环境将会是第一个激活的预置环境，建议使用单个Profile" }
    }

    val rps = profiles.mapNotNull { Enums.getIfPresent(RuntimeProfile::class.java, it.toUpperCase()).orNull() }

    runtimeProfile = if (rps.isEmpty()) {
      logger.warn { "[Environment] - 未找到预置环境（devel, ci, test, prepro, production）" }
      RuntimeProfile.CUSTOM_DEFINED
    } else {
      rps[0]
    }

    this.tempDir = appProperties.tempDir ?: SystemHelper.tempDir()
    logger.info { "[Environment] - 使用临时目录：$tempDir" }

    EnvironmentHandlerHolder.Companion.self = this
    return bean
  }

  class EnvironmentHandlerHolder private constructor() {
    companion object {
      internal lateinit var self: EnvironmentHandler

      fun get(): EnvironmentHandler {
        return self
      }
    }
  }

}

/**
 * 系统运行时环境
 *
 * @author [Dean Zhao](mailto:rcarlosdasilva@qq.com)
 */
enum class RuntimeProfile {
  /**
   * 开发环境
   */
  DEVEL,
  /**
   * 持续集成环境
   */
  CI,
  /**
   * 测试环境
   */
  TEST,
  /**
   * 预生产环境
   */
  PREPRO,
  /**
   * 生产环境
   */
  PRODUCTION,
  /**
   * 自定义
   */
  CUSTOM_DEFINED
}